### PR TITLE
Microcerts fix data inconsistencies across environments

### PR DESCRIPTION
**Pull request by carkod**
## Done

- Fix issues between environments mentioned in https://github.com/canonical-web-and-design/commercial-squad/issues/366 that manifest in https://github.com/canonical-web-and-design/ubuntu.com/pull/10743#issuecomment-957316184
- Refactored API classes into CUBE decorator function and removed reliance on `flask.g` global store
- Removed some variables unused by front-end in the `microcerts.json` endpoint

## QA

- I believe best way to QA this is to use [Morgan's PR](https://github.com/canonical-web-and-design/ubuntu.com/pull/10743#issuecomment-957316184)
- One way to do this, would be to:
```
# Create a new branch for viewing only
git checkout -b carkod-microcerts-fix-envs main
git pull https://github.com/mrgnr/ubuntu.com.git cube-purchase-study-labs
# Pulled Morgan's PR and resolve the conflicts (only 2)
git pull https://github.com/carkod/ubuntu.com.git microcerts-fix-envs
# Pulled this PR
```
- Purchase a course and make sure that the buttons change to "Enrolled" and "Take" now
- Fail a course and check that it's not accessible anymore (Purchase button reappears) issue 359
- Header should activate the correct course you are taking (same as current ubuntu.com)

## Before merging into main branch
Update k8s secrets in prod-comms and demo cluster:
- `BADGR_QA_PASSWORD`
- `CUBE_EDX_QA_CLIENT_ID`
- `CUBE_EDX_CLIENT_QA_SECRET`

Additionally, you can check that passwords have been encoded correctly:
- Connect to VPN
- Get your K8s credentials for demo and prod-comms if you don't have them.
- Do the following commands in demo, prod-comms `--namespace staging` and `--namespace production`
```
kubectl describe secret badgr
kubectl describe secret cube-edx
```
They should include a new `qa-password`, `qa-client-id`, `qa-client-secret` encoded in base64

## Issue / Card

Fixes https://github.com/canonical-web-and-design/commercial-squad/issues/366
Fixes https://github.com/canonical-web-and-design/commercial-squad/issues/359
Fixes https://github.com/canonical-web-and-design/commercial-squad/issues/367
Fixes https://github.com/canonical-web-and-design/commercial-squad/issues/403
Fixes https://github.com/canonical-web-and-design/commercial-squad/issues/410

### CODE REVIEW

File: webapp/advantage/decorators.py [hunk 1]
```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
 
                     return flask.jsonify(message), 401
 
+            badgr_issuer = (
+                BADGR_ISSUER if not test_backend else QA_BADGR_ISSUER
```
Comment (mrgnr) [inline webapp/advantage/decorators.py:172]: Should be `is_test_backend`

File: webapp/advantage/decorators.py [hunk 2]
```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
...
+            )
+            certified_badge = (
+                CERTIFIED_BADGE if not test_backend else QA_CERTIFIED_BADGE,
```
Comment (mrgnr) [inline webapp/advantage/decorators.py:175]: Should be `is_test_backend`

File: webapp/advantage/decorators.py [hunk 3]
```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
...
                 authentication_token=(user_token),
                 token_type=("Macaroon" if user_token else "Bearer"),
                 api_url=get_api_url(is_test_backend),
             )
 
-            if response == "html":
-                g.api.set_is_for_view(True)
+            badgr_api = BadgrAPI(
+                "https://api.eu.badgr.io"
+                if not test_backend
```
Comment (mrgnr) [inline webapp/advantage/decorators.py:188]: Should be `is_test_backend`

File: webapp/advantage/decorators.py [hunk 4]
```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
...
+                else "https://api.test.badgr.com",
+                os.getenv("BAGDR_USER"),
+                os.getenv("BADGR_PASSWORD")
+                if not test_backend
```
Comment (mrgnr) [inline webapp/advantage/decorators.py:192]: Should be `is_test_backend`

File: webapp/advantage/decorators.py [hunk 5]
```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
...
+                else os.getenv("BADGR_QA_PASSWORD"),
+                badgr_session,
+            )
 
-            return func(*args, **kwargs)
+            edx_api = EdxAPI(
+                "https://cube.ubuntu.com"
+                if not test_backend
```
Comment (mrgnr) [inline webapp/advantage/decorators.py:199]: Should be `is_test_backend`

File: webapp/advantage/decorators.py [hunk 6]
```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
...
+                else "https://qa.cube.ubuntu.com",
+                os.getenv("CUBE_EDX_CLIENT_ID")
+                if not test_backend
```
Comment (mrgnr) [inline webapp/advantage/decorators.py:202]: Should be `is_test_backend`

File: webapp/advantage/decorators.py [hunk 7]
```diff
@@ -144,18 +168,57 @@ def decorated_function(*args, **kwargs):
...
+                else os.getenv("CUBE_EDX_QA_CLIENT_ID"),
+                os.getenv("CUBE_EDX_CLIENT_SECRET")
+                if not test_backend
```
Comment (mrgnr) [inline webapp/advantage/decorators.py:205]: Should be `is_test_backend`

**Review by mrgnr — CHANGES_REQUESTED** [review]
Environment switching isn't working properly because `test_backend` is used in some places where we need `is_test_backend`.

File: webapp/cube/api.py [hunk 8]
```diff
@@ -41,6 +41,8 @@ def _authenticate(self):
         data = {"username": self.username, "password": self.password}
 
         response = self.session.post(uri, data=data).json()
+        print("Over here")
```
Comment (jpmartinspt) [inline webapp/cube/api.py:44]: Remove prints

**Review by jpmartinspt — APPROVED** [review]

**Pull request merged**